CES: add option to create CES directly from pods

jshr-w opened the pull request:
This PR makes a change to add an option to create CiliumEndpointSlices directly from pods instead of from CiliumEndpoints. This is a step towards eliminating the need for CEPs when CES are present, which will allow us to reduce pressure created on the kube-apiserver by the CEP CRD at high scale.

Descriptions of changes are in each commit.

```release-note
Add option to create CiliumEndpointSlices directly from pods instead of CiliumEndpoints
```

giorio94 commented on operator/k8s/resources.go:
```diff
@@ -91,3 +94,19 @@ func PodNodeNameIndexFunc(obj interface{}) ([]string, error) {
+	}
+	if node.Spec.Addresses == nil {
+		return []string{}, nil
+	}
```
Nit. Could be dropped, as redundant.

giorio94 commented on operator/k8s/resources.go:
```diff
@@ -91,3 +94,19 @@ func PodNodeNameIndexFunc(obj interface{}) ([]string, error) {
+	if node.Spec.Addresses == nil {
+		return []string{}, nil
+	}
+	indices := []string{}
```
Nit. Could be initialized with a capacity equal to the number of addresses, given that it is known.

qmonnet reviewed (approved):
Looks OK for doc change, thanks!

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -196,13 +216,149 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
+		Name:       pod.GetName(),
+		IdentityID: identityId,
+		Networking: networking,
+		Encryption: ciliumNode[0].Spec.Encryption,
```
Inferring the `Encryption` value from the CN is not guaranteed to be correct in all cases, as it may not match the one of the endpoints. One example I'm currently aware of is when WireGuard node-to-node encryption is enabled, and the given node opted out of node-to-node encryption (i.e., a control plane node). In that case, the CN Encryption value would be set to 0, even though the one of the endpoints should be set to 255.

One possibility in this case could be always setting the wireguard static encryption key when wireguard is enabled, and falling back to the node one only if ipsec is enabled.

Moreover, an update of a CN object (that updates the `Encryption` field) should also enqueue an update for all CES containing endpoints running on that node. Otherwise, the CES will continue referring to a stale key when it gets rotated (which can be the case with IPSec).

jshr-w replied:
Thanks for raising this, I was not familiar with this. Updated

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -196,13 +216,149 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
+}
+
+// Get the identity ID for a given pod.
+func (r *reconciler) getPodIdentityID(pod *slim_corev1.Pod) (int64, error) {
```
Should CID updates enqueue an updates for all CES that reference that identity? Otherwise the risk is that a CES will continue referencing a stale CID that no longer exists (e.g., got manually deleted and then re-created).

sypakine replied:
+1

This needs to reconcile for CID objects. As part of another commit, we need to maintain cid->ces mapping.

giorio94 reviewed (changes requested):
I took a first look at the PR, and left a couple of initial comments inline. I still need to fully digest all the logic (which is fairly new to me) though, as well as look at the tests.

This PR doesn't allow to disable the creation of CEPs yet (due to \[1]), but if that was the case (which I assume being a follow-up step), CES would become incompatible with the clustermesh-apiserver. This component currently only relies on CEP information (which is fine being only a handful of replicas at most). Essentially, this logic \[2] should be extended to work with CES as well, starting either one or the other based on the Cilium configuration. IMO that should be addressed before moving further, to avoid introducing yet another incompatibility between features.

FYI, one of the integration tests failures looks legitimate.

\[1]: https://github.com/cilium/cilium/blob/81666d2a8488c81fba9b51eed35b9c743e778639/pkg/option/config.go#L3242-L3248
\[2]: https://github.com/cilium/cilium/blob/81666d2a8488c81fba9b51eed35b9c743e778639/clustermesh-apiserver/clustermesh/root.go#L392

sypakine commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -347,18 +347,18 @@ func (r *reconciler) allocateCID(cidKey *key.GlobalIdentity) (string, bool, erro
 
-func (r *reconciler) getRelevantLabelsForPod(pod *slim_corev1.Pod) (map[string]string, error) {
-	ns, err := r.getNamespace(pod.Namespace)
+func GetRelevantLabelsForPod(logger *slog.Logger, pod *slim_corev1.Pod, nsStore resource.Store[*slim_corev1.Namespace]) (map[string]string, error) {
```
Can we split this commit into multiple? Refactoring changes are a perfect candidate to be decomposed.

jshr-w replied:
Good point, thanks. Took a stab at splitting it.

pippolo84 commented on operator/pkg/ciliumendpointslice/log.log:
Leftover?

pippolo84 commented on operator/pkg/ciliumendpointslice/controller_test.go:
```diff
@@ -36,51 +39,20 @@ func TestRegisterController(t *testing.T) {
+
+	fakeClient, pod, ciliumEndpointSlice, namespace, ciliumNode, ciliumIdentity, _, hive := initHiveTest(t, true)
+
+	ctx := context.Background()
```
Nit: we can use `t.Context()` (same for all the other occurrences)

pippolo84 commented on operator/pkg/ciliumidentity/testutils/test_utils.go:
```diff
@@ -0,0 +1,86 @@
+func CreateCIDObj(id string, lbs map[string]string) *capi_v2.CiliumIdentity {
+	secLbs := make(map[string]string)
+	for k, v := range lbs {
+		fmt.Printf("Adding label %s=%s\n", k, v)
```
Leftover Printf

pippolo84 reviewed (commented):
Thanks for the PR! :100: 

Spotted a couple of leftovers and left a comment about a minor nit inline, but overall LGTM.

sypakine commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -196,13 +216,149 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
+	for _, podIP := range pod.Status.PodIPs {
+		ip, err := netip.ParseAddr(podIP.IP)
+		if err != nil {
+			logger.Debug("podIP error",
```
This is redundant; the same information will be logged at the call site.

Makes the logger parameter unnecessary.

jshr-w committed:
operator/k8s: index ciliumnodes by ip

Add indexer for CiliumNodes by NodeIP.

This will be required to fetch CiliumNodes by the pod's HostIP when
creating CES directly from pods, without CEPs.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

giorio94 commented on operator/pkg/ciliumendpointslice/testutils/test_utils.go:
Could you please extend the first commit message with the rational behind the change? It currently reads `ciliumidentity: refactor test helpers`, and it is not clear why the tests are being refactored (hence making it difficult for reviewers to figure out what to focus on).

giorio94 commented on operator/pkg/ciliumidentity/testutils/test_utils.go:
```diff
@@ -0,0 +1,84 @@
+	secLbs := make(map[string]string)
+	for k, v := range lbs {
+		secLbs[k] = v
+	}
```
Nit. `secLbs := maps.Clone(lbs)`

giorio94 commented on operator/pkg/ciliumidentity/testutils/test_utils.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+func CreateCIDObj(id string, lbs map[string]string) *capi_v2.CiliumIdentity {
```
Nit. The term `Create` seems to suggest that the object is being created in k8s. Maybe `NewCID` (or simply `CID`) is a bit clearer? Same for the ones below.

giorio94 commented on operator/pkg/ciliumidentity/testutils/test_utils.go:
```diff
@@ -0,0 +1,84 @@
+			Labels: selectedLabels,
+		},
+		SecurityLabels: secLbs,
+	}
```
Nit. Could call the function above rather than duplicating the logic.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -196,13 +219,183 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
+		return "", fmt.Errorf("failed to get CiliumNode from store: %w", err)
+	}
+	return ciliumNode[0].Name, nil
+}
```
Any reason for not simply returning `pod.Spec.NodeName`? That would also allow dropping the extra index altogether.

jshr-w replied:
oh yikes, would have been great to realize this field exists earlier 😅

jshr-w replied:
I also wanted to call out that there ended up being one place here where I still needed this index and so kept it:

When syncing the CES in the local cache, we want to restore the pre-existing state of the cache, so I pulled the node IP from the CoreCEP, and use that to determine the node.
https://github.com/cilium/cilium/blob/a93411f616717bc109a9fc9b6645f723352f6c0c/operator/pkg/ciliumendpointslice/endpointslice.go#L483

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -178,6 +403,26 @@ func NewCEPName(name, ns string) CEPName {
+
+func NewCID(cid string) CID {
+	return CID(cid)
+}
```
Nit. These functions (and likely the ones below) look redundant.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -20,11 +23,23 @@ var (
-	UpdateCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns string) []CESKey
-	RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns string) CESKey
+	AddPodMapping(pod *slim_corev1.Pod, nodeName string, cidKey *key.GlobalIdentity)
+	UpdatePodMapping(pod *slim_corev1.Pod, nodeName string, cid *cilium_v2.CiliumIdentity) []CESKey
```
The naming of these two functions look confusing to me, as they do actually perform different tasks under the hood. AFAIU, `AddPodMapping` only adds the pod to the internal cache, while `UpdatePodMapping` does that and additional operations. How about keeping `AddPodMapping` for the common pieces, and moving the extra bits to another dedicated function with a clearer name?

jshr-w replied:
I've ended up splitting this into
- AddPodMapping
- UpsertPodWithIdentity
- upsertPodIntoCES for the common pieces

Does this feel better to you? I've updated the comments as well

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -6,23 +6,46 @@ package ciliumendpointslice
+
+// To maintain compatibility with duplicate CIDs.
+type SecIDs struct {
+	ids  map[CID]struct{}
```
Nit. I would suggest using `sets.Set[CID]` for more idiomatic set operations. Same for the other occurrences.

pippolo84 commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -281,7 +281,7 @@ func (r *reconciler) cidIsUsedInCEPOrCES(cidName string) bool {
 // 2. CID doesn't exist: Create CID.
 func (r *reconciler) allocateCIDForPod(pod *slim_corev1.Pod) error {
-	k8sLabels, err := r.getRelevantLabelsForPod(pod)
+	k8sLabels, err := GetRelevantLabelsForPod(r.logger, pod, r.nsStore)
```
I don't think the changes in this file belongs to this commit, I suggest to move them in a commit of their own.

pippolo84 commented on operator/k8s/resources.go:
```diff
@@ -91,3 +94,16 @@ func PodNodeNameIndexFunc(obj any) ([]string, error) {
+	if !ok {
+		return []string{}, fmt.Errorf("expected *cilium_v2.CiliumNode, got %T", obj)
+	}
+	indices := make([]string, 0, len(node.Spec.Addresses))
```
Do we really need to consider all the node addresses or is it ok to exclude some types (like `NodeCiliumInternalIP`)?

jshr-w replied:
Oh, this is interesting. For this PR, I am only using the index to get nodes using a pod's HostIP.

https://github.com/cilium/cilium/blob/f34af0d7f042e18aa7236376692d07c906358ccf/operator/pkg/ciliumendpointslice/reconciler.go#L312

https://github.com/cilium/cilium/blob/f34af0d7f042e18aa7236376692d07c906358ccf/operator/pkg/ciliumendpointslice/reconciler.go#L324

And looking [here](https://github.com/kubernetes/kubernetes/blob/a3097010faac734fb4956dbc91ae9034d0a9f840/pkg/util/node/node.go#L64-L65), it seems like a pod's HostIP is guaranteed to be of either `NodeExternalIP` or 
`NodeInternalIP` (which makes sense, I mainly wanted to make sure I'm not missing a case). It seems like it should be okay to only index those two types of node addresses? Updated to reflect this.

pippolo84 replied:
> It seems like it should be okay to only index those two types of node addresses? Updated to reflect this.

Agree. This should give us a leaner index and we can rework it later if there are other consumers that need all the addresses :+1:

pippolo84 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,180 @@ type CESData struct {
+
+	c.nodeNameToEncryptionKey[nodeName] = encryptionKey
+	if _, ok := c.nodeNameToCEPNameSet[nodeName]; !ok {
+		c.nodeNameToCEPNameSet[nodeName] = make(map[CEPName]struct{})
```
if there is no CEP on `nodeName` shouldn't be enough to `return nil` here?

pippolo84 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,180 @@ type CESData struct {
+	defer c.mutex.Unlock()
+
+	if _, ok := c.globalIdLabelsToCIDSet[gidLabels]; !ok {
+		return []CESKey{}
```
Is there a reason to return an empty slice instead of `nil`? Same below at line 211

pippolo84 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -62,17 +65,57 @@ func (c *Controller) initializeQueue() {
+
+	touchedCESs := c.manager.UpdatePodMapping(pod, node, pCid)
 	c.enqueueCESReconciliation(touchedCESs)
+	return
```
Nit: this `return` is unneeded

pippolo84 reviewed (commented):
Thanks for the second iteration!

Took another quick look, I'll do another pass later (need more time to look at all the changes in the last commit).

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -118,9 +150,58 @@ func (c *cesManager) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns stri
+
+func (c *cesManager) UpdateNamespaceMapping(ns *slim_corev1.Namespace) []CESKey {
+	return c.mapping.getCESInNs(ns.GetName())
+}
```
The name of this function is misleading, as no update appears to be actually performed.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -225,8 +286,19 @@ func (c *Controller) runCiliumNodesUpdater(ctx context.Context) error {
 	for event := range c.ciliumNodes.Events(ctx) {
+		switch event.Kind {
+		case resource.Upsert:
+			c.logger.Debug("Got Upsert CiliumNode event", logfields.NodeName, event.Key.String())
```
Nit. `String()` should not be called on log arguments, as slog already does that internally only if the line is actually going to be emitted.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -118,9 +150,58 @@ func (c *cesManager) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns stri
+}
+
+func (c *cesManager) RemoveNodeMapping(node *cilium_v2.CiliumNode) []CESKey {
+	if c.mapping.hasNode(NewNodeName(node.Name)) {
```
Nit. The `hasNode` check could be dropped, as `deleteNode` is idempotent. The pre-check could be also potentially racy given that both `hasNode` and `deleteNode` acquire the lock independently, although that is guaranteed not to happen because only one node event is processed at any time.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -6,23 +6,46 @@ package ciliumendpointslice
 	cesData             map[CESName]CESData
+
+	// nodeNameToCEPNameSet is used to map Node name to all CiliumEndpoints on the node
+	nodeNameToCEPNameSet    map[NodeName]map[CEPName]struct{}
```
How about grouping the information keyed by the same type into a struct? That would make the access pattern easier to reason about. For instance, something like (names TBD):

```go
type NodeInfo struct {
	CEPs   sets.Set[CEPName]
	EncKey EncryptionKey
}

type CESToCEPMapping struct {
    ...
    nodeInfo map[NodeName]NodeInfo
    ...
}
```

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -62,17 +65,56 @@ func (c *Controller) initializeQueue() {
+			c.logger.Info("could not get labels for pod",
+				logfields.K8sPodName, pod.Name,
+				logfields.Error, err)
+			return
```
I assume this should return an error to eventually retrigger a reconciliation, in case the namespace is not found (i.e., the informer is lagging behind). Otherwise, that pod will never receive a CES entry.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -62,17 +65,56 @@ func (c *Controller) initializeQueue() {
+		return
+	}
+
+	_, err := GetPodEndpointNetworking(pod)
```
Any reason for not storing this (and the other `*cilium_v2a1.CoreCiliumEndpoint` bits) directly here, but re-computing them again during the actual reconciliation process?

jshr-w replied:
The main thing I was thinking was that it seems like it should be possible for some of these fields to change between the time this function is called and the CES is reconciled, depending on the queuing delay; so it could reduce the number of CES updates made to the apiserver by getting the most up-to-date data. Does that seem reasonable to you?

giorio94 replied:
Yep, absolutely. Could you maybe add a comment, for future reference?

giorio94 reviewed (commented):
Thanks! Took an initial second look as well, and left a few comments inline. I still need to fully digest and review the last commit though, and I completely skipped the tests for the moment.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -62,17 +65,56 @@ func (c *Controller) initializeQueue() {
+	if err != nil {
+		c.logger.Info("could not get endpointnetworking for pod",
+			logfields.K8sPodName, pod.Name,
+			logfields.Error, err)
```
Does this need to be an info-level log? It would be probably logged for every pod during startup.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
+	if _, ok := c.nodeNameToCEPNameSet[nodeName]; !ok {
+		c.nodeNameToCEPNameSet[nodeName] = make(map[CEPName]struct{})
+		return nil
+	}
```
Is this branch needed at all? All other users of `nodeNameToCEPNameSet` would still need to initialize the map if nil to account for different processing order, so explicitly initializing it here doesn't add value.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
 	defer c.mutex.Unlock()
+
+	if node, ok := c.cepNameToNodeName[cepName]; ok {
+		if _, ok := c.nodeNameToCEPNameSet[node]; ok {
```
This if check doesn't look necessary.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
+	if _, ok := c.globalIdLabelsToCIDSet[gidLabels]; !ok {
+		return nil
+	}
+	secId := c.globalIdLabelsToCIDSet[gidLabels]
```
Nit. Should be merged together.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
+	secId := c.globalIdLabelsToCIDSet[gidLabels]
+	if _, ok := secId.ids[cid]; !ok {
+		return nil
+	}
```
What is the indented behavior in case any endpoint still exists and references the last identity that is being deleted?

jshr-w replied:
I'm thinking that each resource type being deleted clears up its own state. So if the CID is deleted first, the CESs affected will remove references to it, and then the pod deletion would update the CES containing it.
Actually, I may be misinterpreting this question 🤔

giorio94 replied:
My cache is a bit stale here, but I *think* I was referring to the fact that once the identity gets deleted, the endpoint will start pointing to an identity that no longer exists, and further updates will likely fail due to the impossibility of retrieving the identity. This looks a corner case to me, and this handling seems totally reasonable as long as it doesn't affect the reconciliation of the other endpoints. So basically just making sure that everything will eventually converge again to a normal state.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
+	for cepName := range secId.ceps {
+		cesName := c.cepNameToCESName[cepName]
+		cesKeys = append(cesKeys, NewCESKey(cesName.string(), c.cesData[cesName].ns))
+	}
```
This logic could be extracted into a separate helper function (passing the set of CEPs as parameter), and reused for all occurrences.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -118,9 +150,58 @@ func (c *cesManager) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns stri
+		}
+	} else {
+		return c.mapping.insertNode(name, newKey)
+	}
```
This logic could be implemented directly by `insertNode`, to abstract it and ensure that they are all executed under the same lock (even though it doesn't really matter here).

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -20,11 +23,23 @@ var (
 // operations is an interface to all operations that a CES manager can perform.
 type operations interface {
```
Is there any reason for having this interface? There is currently a single implementation, and the number of methods is significantly high, both indications that it may be an unnecessary abstraction.

jshr-w replied:
I think it is just a leftover from having FCFS and Identity-based batching. Added a commit to remove the interface.

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -6,23 +6,46 @@ package ciliumendpointslice
 // CESToCEPMapping is used to map Cilium Endpoints to CiliumEndpointSlices and
 // retrieving all the Cilium Endpoints mapped to the given CiliumEndpointSlice.
 // This map is protected by lock for consistent and concurrent access.
 type CESToCEPMapping struct {
```
Should this file/type be renamed? Its purpose is being significantly extended compared to the previous CES to CEP mapping.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -72,42 +87,59 @@ func (c *cesManager) createCES(name, ns string) CESName {
-	// check the given cep is already exists in any of the CES.
-	// if yes, Update a ces with the given cep object.
+
+	// check the given pod's corecep is already exists in any of the CES.
```
Nit. `s/is already exists/already exists/`

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -72,42 +87,59 @@ func (c *cesManager) createCES(name, ns string) CESName {
+
+// For a pod, return the associated CEP name and CES name (if none already exist,
+// create a new CES)
+func (c *cesManager) addCESForPod(pod *slim_corev1.Pod) (CEPName, CESName) {
```
Would `upsertPodIntoCES` be a better name?

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -118,9 +150,58 @@ func (c *cesManager) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns stri
-		return NewCESKey(cesName.string(), ns)
+		return []CESKey{NewCESKey(cesName.string(), pod.Namespace)}
+	}
+	return []CESKey{CESKey(resource.Key{})}
```
Shouldn't this be `return nil`?

giorio94 commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -35,28 +58,181 @@ type CESData struct {
+}
+
+// Insert a CID into the cache and return the CES which need to be reconciled
+func (c *CESToCEPMapping) insertCID(cid CID, gidLabels string) []CESKey {
```
CIDs should not be updated during normal operation, but this logic would most likely leave leftover entries behind if that were to happen.

jshr-w replied:
Ahh I didn't think this was possible -- I've added some logic to clean up state if the labels associated with a CID change (assuming this is what you are referencing here)

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -118,9 +150,58 @@ func (c *cesManager) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns stri
+
+func (c *cesManager) RemoveIdentityMapping(id *cilium_v2.CiliumIdentity) []CESKey {
+	cidName, gidLabels := cidToGidLabels(id)
+	if c.mapping.hasCID(cidName, gidLabels) {
```
This check looks redundant.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -142,13 +223,25 @@ func (c *cesManager) getLargestAvailableCESForNamespace(ns string) CESName {
-	c.mapping.insertCEP(cepName, ces)
+func (c *cesManager) initializeMappingNode(node *cilium_v2.CiliumNode) {
+	c.mapping.insertNode(NewNodeName(node.Name), NewEncryptionKey(node.Spec.Encryption.Key))
+}
```
What is the difference compared to `UpdateNodeMapping`?

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -142,13 +223,25 @@ func (c *cesManager) getLargestAvailableCESForNamespace(ns string) CESName {
+
+func (c *cesManager) initializeMappingForCID(cid CID, gidLabels string) {
+	c.mapping.insertCID(cid, gidLabels)
+}
```
Why is this necessary opposed to directly calling `UpdateIdentityMapping`?

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -142,13 +223,25 @@ func (c *cesManager) getLargestAvailableCESForNamespace(ns string) CESName {
+// This function is called when we are syncing the cesStore with the manager cache.
+// Therefore we only sync pods that are already mapped to a CES.
+func (c *cesManager) initializeMappingPodToNode(pod *slim_corev1.Pod, ns string, nodeName string, ces CESName, cid CID, gidLabels string) {
+	if _, ok := c.mapping.cesNameToCEPNameSet[ces]; !ok {
```
The access should ideally be protected by a lock, even though no concurrency should still exist at this point. Even better, couldn't the check be moved directly into `insertCEP` to avoid leaking details here?

jshr-w replied:
This makes sense. I wonder whether it might make more sense in this context to just check if a CES has been initialized for the pod before directly calling insertCEP?

giorio94 replied:
Do you mean performing the check directly in the caller of this function? I see that you already have the `cepToCes` map there, so doing it there makes indeed sense to me.

giorio94 commented on operator/cmd/flags.go:
```diff
@@ -294,6 +294,12 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
+	option.BindEnv(vp, option.EnableWireguard)
+
+	flags.Bool(option.EnableIPSecName, defaults.EnableIPSec, "Enable IPsec support")
+	option.BindEnv(vp, option.EnableIPSecName)
```
New flags should be registered via hive, and not added to the global config object.

jshr-w replied:
I tried doing this, please let me know if I misunderstood where this was supposed to be done

giorio94 replied:
I quickly skimmed through the PR, but I don't see where the flags are defined now. What I meant is essentially having a `Config` object like [this](https://github.com/cilium/cilium/blob/3b282690300626fa3bf8682d748dab379f582765/pkg/clustermesh/common/config.go#L21-L32) (with the `Flags` method), and then registering it into hive as [via `cell.Config`](https://github.com/cilium/cilium/blob/3b282690300626fa3bf8682d748dab379f582765/pkg/clustermesh/cell.go#L38).

giorio94 reviewed (commented):
A few more comments inline.

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -23,42 +27,201 @@ type CESData struct {
 	// and the namespace associated with it
 	cesNameToCESData map[CESName]*CESData
+
+	// nodeNameToNodeData is used to map node name to all CiliumEndpoints on the node
```
The number of these internal mappings we are maintaining gives me pause.

Thinking back to the operator's [Allocator leaks](https://github.com/cilium/cilium/pull/38196/commits/6b4e55b6cbe3541d5db085a847e1f2b7cc8c0d0c) makes me worried about the observability/supportability of this change.

I feel like StateDB would address my concerns. I realize that would be a major refactor, but this change overhauls cilium managed identities so there would be no middle-ground between agent-mode and full operator-mode (this change).

jshr-w replied:
Hi @sypakine, definitely see where you are coming from, there are a ton of edge cases in extending the existing design. Could you elaborate a bit on what you are referring to by StateDB here? Or, I'm familiar with StateDB, but what do we want to store in it? I may be missing some context here -- given context, I would be interested in this refactor to make this change safer.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -196,13 +218,181 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
+	if len(storeCIDs) == 0 {
+		return nil, fmt.Errorf("CID store is empty")
+	}
+	return storeCIDs[0], nil
```
How stable is the order in which entries are returned from the store guaranteed to be? Otherwise, there's the risk of significant churn if a set of labels is associated with two or more identities, depending on which one gets picked at any round.

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -85,15 +85,19 @@ func (c *CESToCEPMapping) countCEPs() int {
 	defer c.mutex.RUnlock()
-	ceps := make([]CEPName, 0, c.cesNameToCEPNameSet[ces].Len())
-	for cep := range c.cesNameToCEPNameSet[ces] {
+	ceps := make([]CEPName, 0, c.cesNameToCESData[ces].ceps.Len())
```
return sets.List(c.cesNameToCESData[ces].ceps)

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -103,51 +107,52 @@ func (c *CESToCEPMapping) getCEPsInCES(ces CESName) []CEPName {
 	defer c.mutex.RUnlock()
-	data := c.cesData[name]
-	return data
+	if cesData, ok := c.cesNameToCESData[name]; ok {
```
return c.cesNameToCESData[name]

Miss will return the zero-value of the type (empty for string)

sypakine commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -157,9 +156,9 @@ func (r *reconciler) reconcileCESUpdate(cesName CESName, cesObj *cilium_v2a1.Cil
-	data := r.cesManager.getCESData(cesName)
-	if updatedCES.Namespace != data.ns {
-		updatedCES.Namespace = data.ns
+	data := r.cesManager.getCESNamespace(cesName)
```
rename "data" to "ns" or "namespace"

sypakine commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -22,6 +22,7 @@ import (
 
 	"github.com/cilium/cilium/operator/k8s"
 	cestest "github.com/cilium/cilium/operator/pkg/ciliumendpointslice/testutils"
+	tu "github.com/cilium/cilium/operator/pkg/ciliumidentity/testutils"
```
"identityutils" or something equally meaningful

sypakine commented on operator/pkg/ciliumidentity/testutils/test_utils.go:
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package testutils
```
identityutils or something equally meaningful.

sypakine commented on operator/pkg/ciliumidentity/reconciler.go:
```diff
@@ -345,18 +345,18 @@ func (r *reconciler) allocateCID(cidKey *key.GlobalIdentity) (string, bool, erro
 	return allocatedID.String(), true, nil
 }
 
```
nit: should include a comment for public functions.

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -101,7 +103,7 @@ func (c *CESToCEPMapping) getCEPsInCES(ces CESName) []CEPName {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
-	c.cesNameToCEPNameSet[cesName] = make(map[CEPName]struct{})
+	c.cesNameToCEPNameSet[cesName] = make(sets.Set[CEPName])
```
`sets.New[CEPName]()`

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type CEPName resource.Key
```
`CEPKey` for consistency

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+
+// NodeData contains information about the node; the set of coreceps on
+// the node and the known encryption key associated with the node.
+type NodeData struct {
```
Include the `NodeName`

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+type CEPData struct {
+	ces    CESName
+	node   NodeName
+	labels string
```
Should this be an aliased type? That would be more consistent with the rest of the implementation

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+	cepNameToData map[CEPName]*CEPData
+	// cesNameToData is used to map CiliumEndpointSlice name to all CiliumEndpoints names it contains
+	// and the namespace associated with it
+	cesNameToData map[CESName]*CESData
```
consider: cesData

Similarly for other `...ToData` maps

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+// Handles the cases where the CEP node or labels may have changed.
+func (c *CESCache) updateCEPInCache(cepName CEPName, nodeName NodeName, gidLabels string, cesName CESName) {
+	if cepData, ok := c.cepNameToData[cepName]; ok {
+		// If CEP was previously mapped to different CES, clear.
```
consider: 

func (c *CESCache) clearStaleState(newCEPName CEPName, newNodeName NodeName, newGIDLabels string, newCESName CESName)

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+	c.cesNameToData[cesName].ceps.Insert(cepName)
+
+	if _, ok := c.nodeNameToData[nodeName]; !ok {
+		c.nodeNameToData[nodeName] = &NodeData{
```
consider: constructor when data must be initialized for an object (e.g. a map)

```
func NewNodeData() NodeData {
  return &NodeData{
    ceps: sets.New[CEPName](),
  }
}
```

applies for most/all of the `xData` types

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+	return nil
+}
+
+// Insert a CID into the cache and return the CES which need to be reconciled
```
"... and return all CES which ..."

sypakine commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -0,0 +1,492 @@
+			ceps: sets.New[CEPName](),
+		}
+	}
+	if c.globalIdLabelsToCIDSet[gidLabels].selectedID == "" {
```
consider: introduce a helper function to improve readability. Something like:

```
// SetSelectedID will update the selectedID if it is empty.
func (s *CIDSet) SetSelectedID(newCID CID, override bool) {
  if override {
    s.selectedID = newCID
    return
  }
  s.selectedID = cmp.Or(s.selectedID, newCID)
}

...

c.globalIdLabelsToCIDSet[gidLabels].SetSelectedID(cid, false)
```

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap.go:
```diff
@@ -11,6 +11,10 @@ import (
 	"github.com/cilium/cilium/pkg/lock"
 )
 
+type Cescacher interface {
```
CESCacher

sypakine commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -203,13 +210,17 @@ func (r *reconciler) reconcileCESDelete(ces *cilium_v2a1.CiliumEndpointSlice) (e
-		return k8s.ConvertCEPToCoreCEP(cepObj)
+	if r.cesWithoutCeps {
+		return nil
+	} else {
```
Doesn't need to be within an else block.

sypakine commented on operator/pkg/ciliumendpointslice/ceptocesmap_test.go:
```diff
@@ -9,7 +9,7 @@ import (
 )
 
-func TestCepToCESCounts(t *testing.T) {
+func TestCepToCESCountsDefault(t *testing.T) {
```
CEP

sypakine commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -17,9 +18,10 @@ import (
 
 	"github.com/cilium/cilium/pkg/k8s"
-	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	capi_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
nit: can we not do the package import renaming in the same commit? With so many changes, we should keep commits tightly scoped.

jshr-w committed:
wireguard: export config for enable-wireguard

Export WireGuard --enable-wireguard flag for use in Operator.
CES controller requires --enable-wireguard to set the encryption of the
CES.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

giorio94 commented on pkg/datapath/linux/ipsec/cell.go:
```diff
@@ -22,9 +22,9 @@ var Cell = cell.Module(
-	cell.ProvidePrivate(buildConfigFrom),
+	cell.Config(DefaultUserConfig),
+	cell.Provide(newIPsecAgent, NewIPsecConfig),
+	cell.ProvidePrivate(BuildConfigFrom),
```
AFAIU the reason why you are doing this is that you basically need to know from the operator whether IPSec/Wireguard are enabled. However, this change (and subsequently registering the flags on the operator) has the downside of adding quite a few unnecessary flags to the operator, which don't make sense in that context. 

I'd suggest creating a separate dedicated cell (e.g., `OperatorCell`), which just registers the `EnableIPsec` flag (and symmetrically `EnableWireguard` in the other one). In order to avoid duplication, you could then have the current `UserConfig` struct embed the new one (with the `mapstructure:",squash"` tag, as otherwise it doesn't properly work when registered via `cell.Config`).

/cc @smagnani96 as I know that your intention was to keep these private.

jshr-w replied:
Thank you, this was helpful to understand how to get the Hive configs to work. I was able to get only the necessary flag registered in the operator.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -33,12 +40,19 @@ type cesManager struct {
+		cesManager.cache = newCESCache()
+	} else {
+		cesManager.mapping = newCESToCEPMapping()
+	}
```
The logic looks completely different depending on the mode of operation, as these two fields are alternative to each other. Looking at the methods below, there seems to be very little in common as well, with the vast majority which are intended to be exclusively used either in one case or the other, and the few common ones basically adopting two separate code paths. I think there's a clear opportunity for completely splitting the logic into two separate structs, which IMO would significantly simplify the current review process, as well as future maintenance. It is very difficult otherwise for people without the full context to figure out which method is specific to what, as well as there's significant risk of accidentally accessing variables that are not initialized in that context.

jshr-w replied:
I've separated this out into two separate controller/manager/cache implementations now.  Thank you for the pointers, they were super helpful to figure out how to break this down. I'm hoping this will be easier to review at this point :)

jshr-w replied:
I did not end up moving to the Hive script tests; I was working through it, but it required some significant changes, and this PR is already quite big so I stuck with the Hive test used by the controller for now; I'll probably try to look into adding the tests as script tests separately after these changes.

giorio94 reviewed (changes requested):
I've started looking at the new version, but I'm personally finding it fairly difficult to review. 

Overall, I find the usage of the `CESWithoutCEPs` boolean to split the two code paths to be confusing, because the logic in the two cases is significantly different and with relatively little pieces in common (as also shown by the number of struct fields and methods that used only in one of the two code paths). 

I think that two totally separate implementations (reusing common pieces when appropriate, using interfaces and so on) would make the intention way clearer, helping both the current review phase but also subsequent maintenance and extension.

Splitting the changes into smaller commits would also help quite a lot IMO, if complemented by detailed commit messages providing the rationale for any given subset of changes (in addition to the high level goal of the entire feature), allowing for a more iterative process. The only requirement is that the individual commits compile and that tests can pass, but it is totally fine that the feature does not fully work, or that the flag is not fully respected until the very end.

In this respect, I typically find useful to start from the introduction of the necessary building blocks, each in an individual commit (e.g., configs, new data structures, ...), and complemented by unit tests as appropriate. Then, later commits in the series put everything together, hooking the different building blocks into place. In case of large changes like this, a complementary strategy is to focus on a sub-functionality at a time, e.g., reconciling node changes, and implementing the different pieces necessary to achieve that in a dedicated commit.

Edit. As for tests, I'd suggest checking out the [hive script framework](https://docs.cilium.io/en/stable/contributing/development/hive/#testing-with-hive-script), which simplifies writing integration tests.

jshr-w committed:
ces: rename package

Rename v2alpha1 package correctly, in preparation to import the v2
package.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ipsec: export config for enable-ipsec

Export IPSec --enable-ipsec flag for use in Operator.

CES Controller requires --enable-ipsec in order to set the encryption of
each CES.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add new controller mode

Add config option for `ces-controller-mode`. CES Controller can be
allowed to run with 'default' mode, as it functions today, or the new
mode 'slim', where CES completely replace CEPs in the cluster and CEPs
are no longer created.

When in default mode, the DefaultController is used, which watches CEPs
and creates CES based off those changes.
When in slim mode, the new SlimController will be used instead.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: use a cesmanager interface

The CES Manager must behave differently in 'default' and 'slim' mode, as
we are required to store different state. The default mode will use the
existing ceptocesmap as the datastore, while we introduce a new CESCache
as the datastore for slim mode since we will need to correlate all the
information when we do not rely on CEPs.

We separate out the managers for these modes into a DefaultManager and
SlimManager, which implement the Manager interface. The reconciler depends on the
Manager interface.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add cescache node handling functionality

CES Controller in slim mode will watch CiliumNodes to process updates,
particularly to the EncryptionKey that is stored in CESs.
This requires watching CiliumNodes to handle update and delete. We will
track the set of CEPs on a node, so we know which CESs need to be
reconciled on a node update.

Adds unit tests for node state.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: node update processing flow

The Slim CES Controller process CiliumNode updates with an extra handler
to pass updates to CiliumNode encryption keys to the local CES cache, so
that the affected CESs can be reconciled on changes.

Adds unit tests for CES manager node upsert/delete.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: update cescache cid handling

We extend the CESCache to handle updates to CIDs. We will need to
associate pods with their identities for creating CES.

The SecIds data structure contains a selected ID, which is used if there
are multiple ids associated with a label to minimize churn.

Adds unit test for cid state.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: cid update processing flow

Add flow where controller will process CID updates, update the relevant
mappings in the manager and local CES cache.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add pod state to CESCache

Add store for pod state to CESCache, along with necessary helper
functions.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: handle namespace updates

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: reconciler can be slim or default

We split the reconciler into the default and slim, to pass the
appropriate params depending on the configuration initialized.

Reconciler requires two new interfaces to handle the default/slim modes,
on reconciliation of CES.

The slim reconciler will reconcile based on pod state and local CES
cache state instead of getting data from the CEP.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add pod update/delete flow

When a pod is added/updated or deleted, we insert it into the cache. We
only attempt to add it to the CES cache if we have sufficient state,
otherwise we wait for the necessary info to be ready on the pod.
We will add the pod to the cache whether or not it has been assigned an
identity -- in the case it has not, we will receive an update when the
identity is created and associate it with the pod later.

Add unit tests for cescache and for slim manager flow.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add slim mode CES controller initialization

On initialization of the operator and CES controller, we sync CES from
the store to our local cache before the controller runs. To do this, we
read the existing store state and create the existing CES in our local
cache.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: slimcontroller start flow

Add start flow for SlimController to run.

Modify test helpers for slim mode to be tested.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
ces: add unit tests for ces update flow

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

giorio94 commented on pkg/datapath/linux/ipsec/cell.go:
```diff
@@ -28,6 +28,12 @@ var Cell = cell.Module(
+	"ipsec-operator",
+	"Operator IPSec configuration",
+	cell.Config(defaultEnableConfig),
+)
```
Nit. Could be also simplified to (similarly for the wireguard one):

```go
var OperatorCell = cell.Config(defaultEnableConfig)
```

giorio94 commented on pkg/wireguard/agent/cell.go:
```diff
@@ -93,7 +99,19 @@ type Config struct {
 	EncryptNode      bool
 }
 
+var DefaultEnableConfig = EnableConfig{
```
Nit. Should be unexported for consistency.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -17,10 +17,17 @@ var (
 	sequentialLetters = []rune("bcdfghjklmnpqrstvwxyz2456789")
 )
 
+type CesManager interface {
```
Nit. `CESManager` looks more consistent with the rest of the names (or even only `Manager`).

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -31,23 +38,43 @@ type cesManager struct {
-func newCESManager(maxCEPsInCES int, logger *slog.Logger) *cesManager {
-	return &cesManager{
+type slimManager struct {
+	defaultManager
```
Given that not all fields of `defaultManager` are initialized, I'd suggest to directly add the `logger` and `maxCEPsInCES` fields here (or extract them into a separate struct embedded by both, if valuable). Otherwise there's the risk of invoking the methods provided by `defaultManager`, which would panic as `mapping` is not initialized.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -3,11 +3,98 @@
+
+	if _, ok := c.nodeData[nodeName]; !ok {
+		c.nodeData[nodeName] = NewNodeData()
+		c.nodeData[nodeName].key = encryptionKey
```
Nit. Could pass `key` as parameter of `NewNodeData`.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
I'd suggest squashing this commit (`ces: add unit tests for node state `) into the previous one. They are logically related, and that way the linter does not complain that the methods are unused (it is not a blocker though). And similarly for similar occurrences in subsequent commits.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -0,0 +1,61 @@
+
+	// Insert new Nodes in CES cache and check its total count
+	for _, tc := range testCases {
+		t.Run(tc.name, func(*testing.T) {
```
Don't use `t.Run` if the test depends on the global state, as otherwise the sub-tests cannot be run independently. For instance:

```
$ got . --test.run TestCESCacheNodeState --test.v
=== RUN   TestCESCacheNodeState
=== RUN   TestCESCacheNodeState/Insert_Node_-_1
=== RUN   TestCESCacheNodeState/Insert_Node_-_2
=== RUN   TestCESCacheNodeState/Insert_Same_Node_with_New_Key
=== RUN   TestCESCacheNodeState/Insert_Node_-_1#01
=== RUN   TestCESCacheNodeState/Insert_Node_-_2#01
=== RUN   TestCESCacheNodeState/Insert_Same_Node_with_New_Key#01
--- PASS: TestCESCacheNodeState (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_1 (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_2 (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Same_Node_with_New_Key (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_1#01 (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_2#01 (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Same_Node_with_New_Key#01 (0.00s)
PASS

$ got . --test.run TestCESCacheNodeState/Insert_Node_-_2 --test.v
=== RUN   TestCESCacheNodeState
=== RUN   TestCESCacheNodeState/Insert_Node_-_2
=== NAME  TestCESCacheNodeState
    cescache_test.go:44: 
                Error Trace:    /home/marco/Documents/isovalent/cilium/operator/pkg/ciliumendpointslice/cescache_test.go:44
                Error:          Not equal: 
                                expected: 1
                                actual  : 2
                Test:           TestCESCacheNodeState
                Messages:       Number of nodes in cmap should match with Count
=== RUN   TestCESCacheNodeState/Insert_Node_-_2#01
--- FAIL: TestCESCacheNodeState (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_2 (0.00s)
    --- PASS: TestCESCacheNodeState/Insert_Node_-_2#01 (0.00s)
FAIL
FAIL    github.com/cilium/cilium/operator/pkg/ciliumendpointslice       0.854s
FAIL
```

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -179,3 +183,26 @@ func (c *defaultManager) isCEPinCES(cep CEPName, ces CESName) bool {
+func getNodeEndpointEncryptionKey(node *cilium_v2.CiliumNode, ipsecEnabled, wgEnabled bool) int {
+	switch {
+	case wgEnabled:
+		return cast.ToInt(wgtypes.StaticEncryptKey)
```
Nit. `cast.ToInt` looks overkill of a simple numeric cast.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -230,24 +231,68 @@ func (c *Controller) runCiliumEndpointSliceUpdater(ctx context.Context) error {
+
+func runCiliumNodesUpdater(ciliumNodes resource.Resource[*cilium_api_v2.CiliumNode],
+	rateLimit rateLimitConfig, handleEvent func(event resource.Event[*cilium_api_v2.CiliumNode]),
+	logger *slog.Logger, ctx context.Context) error {
```
Nit. The context should be the first parameter of a function \[1]. I'd expect a linter to complain, but apparently we don't have it enabled.

\[1]: https://pkg.go.dev/context

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -230,24 +231,68 @@ func (c *Controller) runCiliumEndpointSliceUpdater(ctx context.Context) error {
+	)
+}
+
+func runCiliumNodesUpdater(ciliumNodes resource.Resource[*cilium_api_v2.CiliumNode],
```
Nit. Could take `*Controller` as receiver to remove most parameters.

giorio94 commented on operator/pkg/ciliumendpointslice/manager_test.go:
```diff
@@ -7,9 +7,13 @@ import (
 	"github.com/cilium/hive/hivetest"
+	"github.com/spf13/cast"
 	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Nit. `metav1`.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -179,3 +183,26 @@ func (c *defaultManager) isCEPinCES(cep CEPName, ces CESName) bool {
+	return c.mapping.deleteNode(NodeName(node.Name))
+}
+
+func getNodeEndpointEncryptionKey(node *cilium_v2.CiliumNode, ipsecEnabled, wgEnabled bool) int {
```
Could directly return `EncryptionKey`, instead of requiring a further cast afterwards.

giorio94 commented on operator/pkg/ciliumendpointslice/manager_test.go:
```diff
@@ -7,9 +7,13 @@ import (
 	"testing"
 
 	"github.com/cilium/hive/hivetest"
+	"github.com/spf13/cast"
```
Same here about the `cast` package.

giorio94 commented on operator/pkg/ciliumendpointslice/manager_test.go:
```diff
@@ -19,6 +23,19 @@ func getCEP(name string, id int64) *capi_v2a1.CoreCiliumEndpoint {
 	}
 }
 
+func getNode(name string, key int) *capi_v2.CiliumNode {
```
Nit. `newNode` looks more appropriate to me.

giorio94 commented on operator/pkg/ciliumendpointslice/manager_test.go:
```diff
@@ -68,3 +85,39 @@ func TestFCFSManager(t *testing.T) {
+		m.UpdateNodeMapping(getNode("n1", 5), true, false)
+		assert.Len(t, m.mapping.nodeData, 1, "Total number of nodes in cache is 1")
+		key, _ := m.mapping.getEncryptionKey("n1")
+		assert.Equal(t, 5, int(key), "Encryption key for node n1 is 5")
```
Nit. `assert.EqualValues` avoids the need of casts.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -4,13 +4,17 @@
 type NodeName string
 type EncryptionKey int
+type CID string
+type Label string
```
Should this be `Labels`, given that it includes multiple labels?

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,51 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+
+	if _, ok := c.globalIdLabelsToCIDSet[gidLabels]; !ok {
+		c.globalIdLabelsToCIDSet[gidLabels] = NewSecIDs()
+	}
```
Nit: retrieving the value into a variable looks a bit more readable to me, and maybe also a bit more efficient depending on how smart the compiler is (same above):

```go
    secIDs := c.globalIdLabelsToCIDSet[gidLabels]
	if secIDs == nil {
		secIDs = NewSecIDs()
		c.globalIdLabelsToCIDSet[gidLabels] = secIDs
	}
    secIDs...
```

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -98,3 +170,54 @@ func (c *CESCache) getEncryptionKey(nodeName NodeName) (EncryptionKey, bool) {
+
+// Clean up globalIdLabelsToCIDSet map, if no label state exists.
+// Caller should hold cache lock.
+func (c *CESCache) cleanLabelsMap(gidLabels Label) {
```
The convention in many parts of Cilium is to have method names suffixed by `Locked` if they expect the caller to hold the lock. I'd also suggest renaming the method to make it explicit that this only acts if the entry is empty.

jshr-w replied:
Got it, this makes sense. I updated the rest of the methods to follow this convention as well.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -98,3 +170,54 @@ func (c *CESCache) getEncryptionKey(nodeName NodeName) (EncryptionKey, bool) {
 }
+
+// Return if given CID is present in cache
+func (c *CESCache) hasCID(cid CID, gidLabels Label) bool {
```
The `gidLabels` parameter looks unused.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -98,3 +170,54 @@ func (c *CESCache) getEncryptionKey(nodeName NodeName) (EncryptionKey, bool) {
+// If findNextCID is true, it will find the next available CID
+// and update selectedID to it, or set to empty if no CIDs are available.
+// Caller should hold cache lock.
+func (s *SecIDs) setSelectedID(newCID CID, findNextCID bool) {
```
Is there any case in which `newCID` is empty, and `findNextCID` is false? Otherwise the latter could be dropped for simplicity.

jshr-w replied:
Nope, I was thinking about being more explicit about it, but this makes sense.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,51 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+	}
+	c.globalIdLabelsToCIDSet[gidLabels].setSelectedID(cid, false)
+	c.globalIdLabelsToCIDSet[gidLabels].ids.Insert(cid)
+	return c.getCESForCEPs(c.globalIdLabelsToCIDSet[gidLabels].ceps)
```
Should one return early in case the selected CID did not change, to avoid performing unnecessary reconciliations?

jshr-w replied:
Good catches on this function, thank you! Updated the flow related to the CESs returned in this function a bit

jshr-w replied:
I have a question here too. For this and the above comments, I modified the setSelectedID function to return whether the set ID was changed. I'm thinking about this scenario:

CID1 was mapped to Labels1.
Now we see CID1 mapped to Labels2, so we clear up the old state. If the set of Labels1 only had one CID mapped to it, it no longer has a selected ID.
This seems possible; at least I am able to manually update a CID, and the agent creates a new CID for that old set of labels on restart.
In that scenario, would it be preferable to store an empty CID or the old CID? I can't tell if I'm overthinking this :)

jshr-w replied:
It does seem that CEPs keep the old CID. I will update the code flow to keep that pattern, I'm thinking it's more appropriate for the CES CID to align with the datapath endpoint?

giorio94 replied:
> In that scenario, would it be preferable to store an empty CID or the old CID?

AFAIK there's no legitimate case in which the labels associated with a CID should change, and that can only happen if the resource is manually modified (agents/operator would instead always create a new CID resource if labels change). I'm personally fine with either approach, as long as it is consistent, and the overall logic doesn't get confused and is able to keep working correctly.

I think it is reasonable to follow the same logic as CEP mode, even though that one may be also mostly a byproduct of the implementation, more than a reasoned choice.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,51 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+		// If the selectedID is the same as the CID being removed, update it to another valid CID, if it exists.
+		if c.globalIdLabelsToCIDSet[oldLabels].selectedID == cid {
+			c.globalIdLabelsToCIDSet[oldLabels].setSelectedID("", true)
+		}
```
I assume one should also return the `CESKey` entries touched here, as they'd also need to be reconciled again.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,51 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+			if secId.selectedID == cid {
+				// If the selectedID is the same as the CID being removed, update it to another valid CID, if it exists.
+				secId.setSelectedID("", true)
+			}
```
Same here about possibly returning early if nothing changed, to avoid enqueuing unnecessary reconciliations.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -59,3 +59,73 @@ func TestCESCacheNodeState(t *testing.T) {
+
+	// Insert new CEPs in cepCache map and check its total count
+	for _, tc := range testCases {
+		t.Run(tc.name, func(*testing.T) {
```
Same here about the problems of subtests that depend on the global state.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -59,3 +59,73 @@ func TestCESCacheNodeState(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(*testing.T) {
+			prevSelectedId, _ := cmap.GetSelectedId(tc.gidLabels)
+			cmap.insertCID(CID(tc.cid), Label(tc.gidLabels))
```
Nit. Unnecessary casts.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -59,3 +59,73 @@ func TestCESCacheNodeState(t *testing.T) {
+		t.Run(tc.name, func(*testing.T) {
+			prevSelectedId, _ := cmap.GetSelectedId(tc.gidLabels)
+			cmap.insertCID(CID(tc.cid), Label(tc.gidLabels))
+			assert.Equal(t, len(cmap.cidToGidLabels), tc.count, "Number of CIDs in cmap should match with Count")
```
Nit. `assert.Len` (the linter should catch this, but apparently it is not for some reason :disappointed:)

jshr-w replied:
Thought the linter would catch me here! I did a pass through all the tests... hopefully I got them all 🤞

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -59,3 +59,73 @@ func TestCESCacheNodeState(t *testing.T) {
 	assert.Empty(t, cmap.nodeData)
 }
+
+func TestCESCacheCIDState(t *testing.T) {
```
It would be good to also test the problematic cases, such as changing the labels associated with a CID, removing one of two CIDs and checking it selects another one, and so on.

jshr-w replied:
Yup makes sense -- I added the latter into this commit. I added a test for labels change in a later commit because it depends on CEPs being present in the map as well in its current functionality.

giorio94 reviewed (changes requested):
Thanks! The commits structure looks great, and I personally find the PR way easier to review. I've started looking into the first half of the commits, and left a bunch of comments inline, which the majority being nits. I'll continue with the rest over the next days.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache_test.go:
```diff
@@ -0,0 +1,57 @@
+	for _, tc := range testCases {
+		cmap.insertNode(tc.nodeName, tc.encryptionKey)
+		assert.Len(t, cmap.nodeData, tc.count, "Number of nodes in cmap should match with Count")
+		nd, ok := cmap.nodeData[tc.nodeName]
```
I'd suggest checking the outputs of `hasNode` and `getEncryptionKey`, instead of directly checking the map, in order to validate the actual APIs.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,65 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+		if cidSet.selectedID == cid {
+			changed := cidSet.setSelectedIDLocked("")
+			if changed {
+				cesToReconcile = append(cesToReconcile, c.getCESForCEPs(cidSet.ceps)...)
```
`s/getCESForCEPs/getCESForCEPsLocked` (and similarly for a couple of other occurrences in this commit)

jshr-w replied:
Hit the wrong commit :(

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,65 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+				cesToReconcile = append(cesToReconcile, c.getCESForCEPs(cidSet.ceps)...)
+			}
+		}
+		c.globalIdLabelsToCIDSet[oldLabels] = cidSet
```
Nit. Unnecessary, as `cidSet` is a pointer.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -74,6 +101,65 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+		secIDs = NewSecIDs()
+		c.globalIdLabelsToCIDSet[gidLabels] = secIDs
+	}
+	changed := c.globalIdLabelsToCIDSet[gidLabels].setSelectedIDLocked(cid)
```
Nit. `s/c.globalIdLabelsToCIDSet[gidLabels]/secIDs` (and similarly for the subsequent occurrences).

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -208,3 +209,19 @@ func getNodeEndpointEncryptionKey(node *cilium_v2.CiliumNode, ipsecEnabled, wgEn
+
+func cidToGidLabels(id *cilium_v2.CiliumIdentity) (CID, Labels) {
+	cidName := id.GetName()
+	cidKey := key.GetCIDKeyFromLabels(id.SecurityLabels, "")
```
Should the second parameter be `labels.LabelSourceK8s`?

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -50,6 +55,11 @@ func NewSecIDs() *SecIDs {
 type CESCache struct {
 	mutex lock.RWMutex
 
+	// cesData is used to map CiliumEndpointSlice name to all CiliumEndpoints names it contains
+	// and the namespace associated with it
+	cesData map[CESName]*CESData
+	// nsData is used to map namespaces to all CiliumEndpointSlices in them
+	nsData map[string]sets.Set[CESName]
```
This does not work correctly if two CESes have the same name in different namespaces. Today CESes are cluster scoped, so that is effectively impossible, but it may be something to keep in mind if that were to change in the future.

jshr-w replied:
This is a fair call out -- I will add this tentatively to the roadmap and address it (I'll keep an eye on the plans to move CES to cluster-scope for the timing of this)

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -160,6 +172,41 @@ func (c *CESCache) deleteCID(cid CID, gidLabels Labels) []CESKey {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	for ces := range c.nsData[ns] {
+		c.cesData[ces].ns = ""
```
Why the namespace name is reset here, rather than deleting the whole object? Wouldn't that risk leaving stale objects behind in the cache? I'm wondering whether more in general this `deleteNs` method needs to exist, or it shouldn't be the responsibility of the callers to delete all CESes in K8s, and consequently call `deleteCES` here.

jshr-w replied:
I do agree -- I think it's better for deleteCES to be called when a CES is empty after a sequence of pod deletes. My concern was about leftover namespace keys, which is why I introduced this method. But I'm realizing it would be better to clean up namespace objects when possible after `deleteCES`.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -160,6 +172,41 @@ func (c *CESCache) deleteCID(cid CID, gidLabels Labels) []CESKey {
+	if _, ok := c.nsData[ns]; !ok {
+		c.nsData[ns] = sets.New[CESName]()
+	}
+	c.nsData[ns].Insert(cesName)
```
This works under the assumption that the namespace of a CES doesn't ever change, and that's not expected to happen, but I'm not sure whether it is actually enforced to never happen.

jshr-w replied:
Hmm I adjusted this function to not make that assumption -- I feel like this shouldn't happen, but the reconciler logic accounts for it so possibly I'm missing where it can happen.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -216,10 +263,83 @@ func (c *CESCache) cleanLabelsMapIfNoStateLocked(gidLabels Labels) {
+	for ces := range c.cesData {
+		cess = append(cess, ces)
+	}
+	return cess
```
Nit. `return slices.Collect(maps.Keys(c.cesData))` (and similarly below)

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -216,10 +263,83 @@ func (c *CESCache) cleanLabelsMapIfNoStateLocked(gidLabels Labels) {
+	return ""
+}
+
 func (cid CID) String() string {
```
Nit. This `String()` function looks out of place.

jshr-w replied:
Fair! This is a remnant of the pattern from the old ceptocesmapping. Adjusted to not need this anymore

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -216,10 +263,83 @@ func (c *CESCache) cleanLabelsMapIfNoStateLocked(gidLabels Labels) {
+// Return if the given CES is present in cache
+func (c *CESCache) hasCESName(cesName CESName) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
```
Commenting here, but this is a more general comment.

All these methods are protected by the lock, which ensures that the underlying data structures are never accessed concurrently. However, their scope is fairly narrow, and to me they look fairly prone to misuse.

For instance, let's imagine the following arbitrary flow:

```go
if c.hasCESName("foo") && c.countCEPsInCES("foo") == 0 {
	c.deleteCES("foo")
}
```

There's no guarantee that when `deleteCES` gets called any of the other properties still hold, as the lock was released and reacquired, and other operations may have occurred in the meanwhile. 

I'm not say that the implementation in the subsequent commits is necessarily affected by this, but I wonder if it wouldn't be more appropriate to drop the lock from CESCache, and move it to the callers instead, to be able to hold it for the entire duration of the operation. Or alternatively introduce self-contained methods here, if that's possible.

jshr-w replied:
I see your point. I think this issue could exist in the existing system as well? For this, I think I find that moving the lock to the SlimManager seems best to me, so that we can lock the entire set of operations on the mapping, so I went with your first suggestion.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -216,10 +263,83 @@ func (c *CESCache) cleanLabelsMapIfNoStateLocked(gidLabels Labels) {
+}
+
+// Return the total number of CESs.
+func (c *CESCache) getCESCount() int {
```
Nit. Methods only intended for use in tests should be moved to test files.

giorio94 commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -82,12 +83,20 @@ func newSlimManager(maxCEPsInCES int, logger *slog.Logger) *slimManager {
+	return createCES(name, ns, c.mapping, c.logger)
+}
+
+func createCES(name, ns string, cacher CESCacher, logger *slog.Logger) CESName {
```
For instance this function is not safe to be called concurrently, because of two separate calls to cacher methods that independently handle locking.

giorio94 commented on operator/pkg/ciliumendpointslice/namespace.go:
```diff
@@ -59,5 +69,13 @@ func (c *Controller) onNamespaceDelete(ns *slimcorev1.Namespace) {
+}
 
+func (c *SlimController) onNamespaceDelete(ns *slimcorev1.Namespace) {
+	c.logger.Debug(fmt.Sprintf("Namespace deleted: %s", ns.Name))
```
Nit. Should use attributes instead of `fmt.Sprintf`.

giorio94 commented on operator/pkg/ciliumendpointslice/namespace.go:
```diff
@@ -59,5 +69,13 @@ func (c *Controller) onNamespaceDelete(ns *slimcorev1.Namespace) {
 
+func (c *SlimController) onNamespaceDelete(ns *slimcorev1.Namespace) {
+	c.logger.Debug(fmt.Sprintf("Namespace deleted: %s", ns.Name))
+	c.manager.RemoveNamespaceMapping(ns)
```
I'm not convinced that this is correct. In theory at this point there should be no longer any CES for that namespace (or possibly be queued for processing), as the corresponding pods have been deleted as well, but to be on the safe side one may enqueue a reconciliation for any of remaining entry.

jshr-w replied:
Resolving, since I've removed this function based on a different comment thread

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -160,6 +172,41 @@ func (c *CESCache) deleteCID(cid CID, gidLabels Labels) []CESKey {
+	defer c.mutex.Unlock()
+
+	if cesData, ok := c.cesData[cesName]; ok {
+		if cesSet, ok := c.nsData[cesData.ns]; ok {
```
Nit. May delete the `nsData` entry if this was the last CES in that namespace.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -105,7 +116,9 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
-		// TODO: Clean up node state in CEPData
+		for cepName := range nodeData.ceps {
+			c.cepData[cepName].node = ""
+		}
```
Why is the node name reset here? Shouldn't it be preserved until the pod disappears?

giorio94 reviewed (changes requested):
Thanks! I took a fresh look at a few more commits, and left some more comments inline.

jshr-w committed:
ces: add cescache ces and ns handling

Allow the local CES datastore to store CES and namespace information.

We also introduce the CESCacher interface, to allow the default and slim
CES managers to handle their own local cache types.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -65,8 +92,90 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+}
+
+// Remove CID from cache and return affected CESs
+func (c *CESCache) deleteCID(cid CID, gidLabels Labels) []CESKey {
```
Should only take the `CID` as parameter, and retrieve the labels from `cidToGidLabels`, to prevent risking issues in case the given `gidLabels` are not in sync with the currently cached ones (looks a bit of a footgun, regardless of whether that's actually possible today).

jshr-w replied:
Fair call out, thanks!

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -65,8 +92,90 @@ func (c *CESCache) deleteNode(nodeName NodeName) []CESKey {
+func (c *CESCache) deleteCID(cid CID, gidLabels Labels) []CESKey {
+	delete(c.cidToGidLabels, cid)
+	if secId, ok := c.globalIdLabelsToCIDSet[gidLabels]; ok {
+		if secId.ids.Has(cid) {
```
Nit. This `Has` check could be likely omitted for simplicity, as always expected to be true.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -195,8 +206,14 @@ func (c *CESCache) deleteCES(cesName CESName) {
 
 // Return CES keys for the given CEPs. Caller must hold the cache lock.
 func (c *CESCache) getCESForCEPs(ceps sets.Set[CEPName]) []CESKey {
```
This function is likely to return duplicate keys in case multiple CEPs are associated with the same CES. I'd expect the callers to not have much problems with that (worth double checking though), but at least logs may easily get confusing:

>  c.logger.Debug("Enqueueing CES (if not empty name)", logfields.CESName, ces.string())

jshr-w replied:
Yeah, I was thinking that when these CES are enqueued multiple times the updates would be batched into 1 because of the sync delay, but it's feasible to just avoid it and not have too many repetitive logs. Moved to using sets to collect the unique CESs updated.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -195,8 +206,14 @@ func (c *CESCache) deleteCES(cesName CESName) {
-	return nil
+	cesKeys := make([]CESKey, 0, ceps.Len())
+	for cepName := range ceps {
+		if cepData, ok := c.cepData[cepName]; ok {
```
Nit. Could use `getCESName` instead of open-coding it.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -214,3 +254,172 @@ func (r *reconciler) getCoreEndpointFromStore(cepName CEPName) *cilium_v2a1.Core
+	scopedLog := r.logger.With(
+		logfields.K8sPodName, pod.GetName(),
+		logfields.K8sNamespace, pod.Namespace,
+	)
```
Nit. Should not use `logger.With` for one-off logs, as expensive in case debug logs are disabled (compared to passing the attributes directly).

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -214,3 +254,172 @@ func (r *reconciler) getCoreEndpointFromStore(cepName CEPName) *cilium_v2a1.Core
+			logfields.Identity, cid,
+			logfields.Error, err,
+		)
+		return 0, fmt.Errorf("failed to parse CID name: %w", err)
```
Nit. It is typically a code smell to log on error and return the error as well, as likely leading to duplicate logs. Only the outermost caller should take care of logging, while the inner functions should just propagate the (possibly enriched) error.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -214,3 +254,172 @@ func (r *reconciler) getCoreEndpointFromStore(cepName CEPName) *cilium_v2a1.Core
+	return networking, nil
+}
+
+func (r *slimReconciler) getEndpointEncryptionKey(pod *slim_corev1.Pod) (int, error) {
```
Any reason for re-implementing this logic, rather than querying the CESCache?

jshr-w replied:
Yes, I was thinking about this kind of scenario:
- Pod A is created on Node A
- CES Controller's informer sees the delete for Node A, so Node A's encryption key is removed from the CESCache and reconciliation is enqueued for CES A (which contains Pod A)
- But we haven't got the delete for Pod A yet (or update for Pod A being rescheduled to a different node yet...) when reconcileCESUpdate is called
- reconcileCESUpdate calls getCoreEndpointFromStore for every ccep in CES A

If we fetch the EncryptionKey from the cache, getCoreEndpointFromStore would return a nil ccep, so the CES update call made to the APIServer removes the pod from its coreceps. Which I think would cause unnecessary churn, if the pod were just being rescheduled and not deleted. 

So, my thinking was that fetching the state here avoids this churn step if encryption is not enabled or Wireguard is enabled. I believe we will still have the same issue for the IPSec case, but it will be reconciled back if necessary when we receive the pod state for rescheduling (or in the case of delete, the cescache will be cleared when we get the pod delete).

Originally I would have liked to match the cache state here, but this scenario made me feel like this would perform at least a bit better. What do you think?

giorio94 replied:
> Which I think would cause unnecessary churn, if the pod were just being rescheduled and not deleted.

Not sure I'm fully following, as a rescheduled pod would be anyways associated with different IP addresses (and most likely a different name), which would cause an update to the CES.

My main concern with not querying the CESCache here is that it is easier to end up in situations in which the state gets out of sync (e.g., the given cep is not tracked in `NodeData`, which may subsequently lead to leftover entries). I'd personally suggest to favor a more straightforward and less error prone approach, even if it may be slightly more suboptimal from the performance point of view in rare corner cases.

giorio94 reviewed (changes requested):
Thanks, I almost got through all commits. A few more minor comments inline, tomorrow I'll try to look at the final three commits.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -214,3 +254,167 @@ func (r *reconciler) getCoreEndpointFromStore(cepName CEPName) *cilium_v2a1.Core
+		return r.convertPodToCoreCEP(podObj)
+	}
+	r.logger.DebugContext(r.context,
+		fmt.Sprintf("Couldn't get Pod from Store (err=%v, exists=%v)", err, exists),
```
Nit. Should provide the extra information via attributes, not `fmt.Sprintf`.

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -94,6 +97,82 @@ func newCESCache() *CESCache {
+	c.cesData[cesName].ceps.Insert(cepName)
+
+	if _, ok := c.nodeData[nodeName]; !ok {
+		c.nodeData[nodeName] = NewNodeData(0)
```
Creating an entry associated with key 0 doesn't look correct to me, and potentially a security issue, because it may cause encryption to be temporarily disabled for the given pod, until the update for the corresponding node is actually observed. It can also lead to stale state, given that there's no guarantee that `deleteNode` will be ever called afterwards.

I'd personally suggest going with a more straightforward approach, returning an error if any expected entry of the cache (node, identity, ...) is not found, and let the caller logic retry later once they have been observed. Mangling with the content of unrelated fields is likely to cause the state to get out of sync.

jshr-w replied:
Agreed on the security issue. I have moved the retry logic to the reconciler. At the CES cache, I went forward with a model where we can see if the key has been set after a node update or not, so that we are able to store node->pod[] mappings, but only reconcile their CESs after the node updates have been received.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -418,3 +421,32 @@ func getProtocolString(p slim_corev1.Protocol) (string, error) {
+		return nil, fmt.Errorf("CID store is empty")
+	}
+	return storeCIDs[0], nil
+}
```
Similarly here, why is it necessary to reimplement this rather than relying on the `CESCache`? 

Querying the identities directly would also bypass the logic to keep stable CIDs in case of duplicate identities, unless I'm missing anything of the big picture.

jshr-w replied:
This function is only called from onPodUpdate(), so this is where we get the initial identity that is put into the cache. It's not called on reconciliation, where we use getPodIdentityIDFromCache instead.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -313,6 +314,65 @@ func (c *SlimController) onIdentityDelete(cid *cilium_api_v2.CiliumIdentity) {
+		// in the ces cache, so it can be associated with the CID once it is
+		// created.
+		c.manager.AddPodMapping(pod, node, cidKey)
+		return err
```
Is it correct/necessary to return an error? I'd expect reconciliation to be triggered when the identity appears, and not require a retry here

giorio94 commented on operator/pkg/ciliumendpointslice/cescache.go:
```diff
@@ -94,6 +97,82 @@ func newCESCache() *CESCache {
+	c.updateCEPInCache(cepName, nodeName, gidLabels, cesName)
+	c.globalIdLabelsToCIDSet[gidLabels].setSelectedID(cid)
+	c.globalIdLabelsToCIDSet[gidLabels].ids.Insert(cid)
+	c.cidToGidLabels[cid] = gidLabels
```
This doesn't look correct to me, because it manually updates the CID/label mappings, bypassing the  `insertCID/deleteCID` methods. Which in turn can lead to stale or out-of-sync state, as the invariants are no longer guaranteed (e.g., in case this method gets called after that the deletion even of a CID has already been seen). I'd suggest to simplify the logic by always looking into the CESCache to retrieve the identity to CID mapping, and simply trigger a retry if not available yet.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -391,6 +393,63 @@ func (c *DefaultController) syncCESsInLocalCache(ctx context.Context) error {
+					logfields.Error, err)
+				continue
+			}
+			c.manager.initializeMappingPodToNode(NewCEPName(cep.Name, ces.Namespace), NodeName(nodeObj[0].Name), CESName(ces.Name), CID(identityid), Labels(labels), EncryptionKey(cep.Encryption.Key))
```
This may leave stale state behind in case a given resource is deleted before subscribing to the event channels, as the corresponding deletion event will never be observed.

A safer approach would be to not rely on the stores for this initialization, but rather perform that inline as part of the normal processing of events. For instance, you could have the cilium node and identity updaters close a channel when their sync event is received. At that point, you could start the endpointslice observer, and do restoration until the sync event for that is received. And finally start the rest of the logic that depends on that. This ensures that any deletion that may occur in between is guaranteed to be eventually observed and processed.

jshr-w replied:
This is a good point. I intend to make these changes, but I'd like to push this to a follow up PR so that I can get a review of the changes involved and also given the fact that the default controller behaves the same way (probably could use a similar approach there as well).

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -196,6 +196,52 @@ func (c *DefaultController) Start(ctx cell.HookContext) error {
+	c.wp.Submit("cilium-pods-updater", c.runCiliumPodsUpdater)
+	c.wp.Submit("cilium-endpoint-slices-updater", c.runCiliumEndpointSliceUpdater)
+	c.wp.Submit("cilium-nodes-updater", c.runCiliumNodesUpdater)
+	c.wp.Submit("cilium-identities-updater", c.runCiliumIdentitiesUpdater)
```
I'd suggest switching all of these to be jobs, for consistency.

jshr-w replied:
done. fwiw this pattern exists in the default controller mode

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler.go:
```diff
@@ -98,6 +98,47 @@ func newDefaultReconciler(
+	ipsecEnabled bool,
+	wgEnabled bool,
+) *slimReconciler {
+	cesStore, _ := ciliumEndpointSlice.Store(ctx)
+	podStore, _ := pod.Store(ctx)
+	cidStore, _ := ciliumIdentity.Store(ctx)
+	ciliumNodeStore, _ := ciliumNode.Store(ctx)
+	namespaceStore, _ := namespace.Store(ctx)
```
The context used here should be the one from the start hook, otherwise it would not be possible to abort the start hook, as this context will never be canceled. You may extract their initialization outside of this function, of add something like the following before invoking this method at the caller site:

```go
	stop := context.AfterFunc(ctx, c.contextCancel)
	defer stop()

    c.manager = newSlimReconciler(...)
    ...
```

jshr-w replied:
I will push this to a follow-up PR as well, given that the default controller behaves similarly and it involves some changes around the behavior that would probably deserve a review.

giorio94 commented on operator/pkg/ciliumendpointslice/controller_test.go:
```diff
@@ -149,3 +152,136 @@ func createCEPandVerifyCESCreated(t *testing.T, fakeClient k8sClient.Clientset,
+		return len(ciliumNodeStore.List()) > 0
+	}, time.Second); err != nil {
+		return false, fmt.Errorf("failed to get CiliumNode: %w", err)
+	}
```
Any specific reason for this check (and the analogous ones below)? Seem redundant to me given that you are testing that the resources you manually created as part of the test exist in the store (which doesn't even mean that they have been processed by the logic under test).

jshr-w replied:
just duplicated existing tests -- updated

giorio94 commented on operator/pkg/ciliumendpointslice/controller_test.go:
```diff
@@ -149,3 +152,136 @@ func createCEPandVerifyCESCreated(t *testing.T, fakeClient k8sClient.Clientset,
+	return err == nil, nil
+}
+
+func initHiveTest(t *testing.T, enableCES, enableCESwithoutCEPs bool) (k8sClient.Clientset, *resource.Resource[*slim_corev1.Pod], *resource.Resource[*cilium_v2a1.CiliumEndpointSlice], *resource.Resource[*slim_corev1.Namespace], *resource.Resource[*cilium_v2.CiliumNode], *resource.Resource[*cilium_v2.CiliumIdentity], *Metrics, *hive.Hive) {
```
Nit. Resource[T] are typically passed by value, not pointer, as they are already an interface.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -258,3 +265,356 @@ func TestReconcileNoop(t *testing.T) {
+	m.mapping.upsertCEP(NewCEPName("pod2", "ns"), CESName("ces2"), "node1", gidB, "2")
+	m.mapping.upsertCEP(NewCEPName("pod3", "ns"), CESName("ces2"), "node1", gidB, "2")
+
+	// ces1 contains cep1 and cep3, but it's mapped to nothing so it should be deleted
```
Nit. The comment doesn't look correct.

giorio94 commented on operator/pkg/ciliumendpointslice/reconciler_test.go:
```diff
@@ -258,3 +265,356 @@ func TestReconcileNoop(t *testing.T) {
 	hive.Stop(tlog, t.Context())
 }
+
+func TestReconcileCreate(t *testing.T) {
```
It would be nice to convert the vast majority of higher level tests into script tests, which would be way easier to read, and allow for much better coverage testing the entire flow E2E. It can be a follow-up item though.

jshr-w replied:
ACK, and this is on my TODO list as a follow-up PR.

giorio94 reviewed (approved):
Sorry for the delay. I took one more look at the PR, and left some more comments inline. Overall, I'm still a bit concerned about the the complexity of the state tracking logic, and the different corner cases in which I feel it may get out of sync, potentially never recovering anymore. 

That said, I won't block this PR any further, as I'll be out for a couple of weeks, in case you want to get it merged before feature freeze. Under the assumption that my main comments get addressed before merging, and that the entire feature is behind a hidden feature flag, which would allow for subsequently further iterating on possible issues without affecting users. As a follow-up, I think that scaffolding some script tests may definitely help to increase confidence in the overall functionality.

rgo3 reviewed (approved):
Only looked at commits 3 and 4 for the ipsec and wireguard config changes. LGTM

pippolo84 reviewed (approved):
✅
